src/hotspot/share/gc/shenandoah/shenandoahEvacOOMHandler.cpp

```diff
@@ -31,22 +31,46 @@
 
 const jint ShenandoahEvacOOMHandler::OOM_MARKER_MASK = 0x80000000;
 
-ShenandoahEvacOOMHandler::ShenandoahEvacOOMHandler() :
-  _threads_in_evac(0) {
+ShenandoahEvacOOMHandler::ShenandoahEvacOOMHandler() {
+  for (int i = 0; i < EVAC_COUNTER_BUCKETS; i++)
+    _threads_in_evac[i].bits = 0;
 }
 
-void ShenandoahEvacOOMHandler::wait_for_no_evac_threads() {
-  while ((Atomic::load_acquire(&_threads_in_evac) & ~OOM_MARKER_MASK) != 0) {
+volatile jint *ShenandoahEvacOOMHandler::threads_in_evac_ptr(Thread* t) {
+  uint64_t key = (uintptr_t)t;
```

Comment:
Maybe put that in a separate hash(Thread*) function? Also, is that a particular documented hash-function?(Related: In Lilliput project, I am working on a different identity-hash-code implementation, and part of it will be a hash-implementation to hash arbitrary pointers to 32 or 64 bit hash, currently using murmur3. Maybe this could be reused for here, when it happens?)

Reply:
It is actually the bit mixing function from MurmurHash3. The particular algorithm doesn't matter too much though - I just couldn't find an existing one in the shared code.